### PR TITLE
fix(Row): types for runtime conditional values

### PR DESCRIPTION
**Pull request by brtbrt**
it's now possible using runtime conditions that makes enable/disbale both `switch` and `checkbox` properties along with a fixed `onPress` property, which in turn will make the component act like a Base one.

- [x] flow to ignore the new utility type
- [x] typescript fix: it depends on the first js-utils PR in order to generate the proper flowjs type and make pass the type verification. check out the [related PR](https://github.com/Telefonica/js-toolbox/pull/57).

About the `ExclusifyUnion` utility type I've placed: that was an idea to make things more clear and easier to write/maintain, not the opposite. If you can't wrap your mind on the concept, or you see it overcomplicates the FlowJS compatibility, let me know and we can rollback the change.

In my opinion it does help where it's really needed (the components themselves) to maintain those types with an open-closed principle that wasn't followed. The utility type itself, once understood for this PR itself, it could be seen as black box.

valid:
```typescript
const Test = () => {
    const v = true;

    return (
        <BoxedRow
            title="hola"
            onPress={() => {
                //  nada
            }}
            checkbox={v ? {defaultValue: true, onChange: (val) => console.log(val)} : undefined}
        />
    );
};
```

valid:
```typescript
const Test = () => {
    const v = true;

    return (
        <BoxedRow
            title="hola"
            onPress={() => {
                //  nada
            }}
        />
    );
};
```

valid:
```typescript
const Test = () => {
    const v = true;

    return (
        <BoxedRow
            title="hola"
            onPress={() => {
                //  nada
            }}
            switch={v ? {defaultValue: true, onChange: (val) => console.log(val)} : undefined}
        />
    );
};
```

valid:
```typescript
        <BoxedRow
            title="hola"
            checkbox={v ? {defaultValue: true, onChange: (val) => console.log(val)} : undefined}
        />
```

INVALID:
```typescript
const Test = () => {
    const v = true;

    return (
        <BoxedRow
            title="hola"
            onPress={() => {
                //  nada
            }}
            checkbox={v ? {defaultValue: true, onChange: (val) => console.log(val)} : undefined}
            switch={v ? {defaultValue: true, onChange: (val) => console.log(val)} : undefined}
        />
    );
};
```

### CODE REVIEW

File: src/list.tsx [hunk 1]
```diff
@@ -258,107 +259,93 @@ type ControlProps = {
...
     | BasicRowContentProps
-    | SwitchRowContentProps
     | RadioRowContentProps
-    | CheckboxRowContentProps
     | HrefRowContentProps
     | ToRowContentProps
-    | OnPressRowContentProps;
+>;
+
+function areRadioRowContentProps(obj: any): obj is RadioRowContentProps {
```
Comment (atabel) [inline src/list.tsx:312]: please, follow coding style

File: src/utils/utility-types.d.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+type AllKeys<T> = T extends unknown ? keyof T : never;
```
Comment (atabel) [inline src/utils/utility-types.d.ts:1]: I'd use the `src/utils/types.tsx` file.

Also, you don't need any change in `js-toolbox` if you follow [the same approach we did with the `RendersElement`](https://github.com/Telefonica/mistica-web/pull/413) utility type: you can define an override for flow inside the `flow-overrides/` folder.

File: src/list.tsx [hunk 3]
```diff
@@ -24,6 +24,7 @@ import {getPrefixedDataAttributes} from './utils/dom';
```
Comment (atabel) [inline src/list.tsx:24]: please, can you define some new type tests in `list-tipe-tests.tsx` to cover the cases this PR is fixing?

File: src/list.tsx [hunk 4]
```diff
@@ -258,107 +259,93 @@ type ControlProps = {
...
     onPress: () => void;
-    href?: undefined;
-    to?: undefined;
-    right?: Right;
 }
 
-type RowContentProps =
+type RowContentProps = ExclusifyUnion<
+    | (Omit<SwitchRowContentProps, 'switch'> & {switch: ControlProps | undefined})
+    | (Omit<CheckboxRowContentProps, 'checkbox'> & {checkbox: ControlProps | undefined})
```
Comment (atabel) [inline src/list.tsx:304]: can you explain this? why is the Omit needed if you're re-defining the `switch`/`checkbox` prop with the same type? I don't get it

File: src/list.tsx [hunk 5]
```diff
@@ -258,107 +259,93 @@ type ControlProps = {
...
 
-interface OnPressRowContentProps extends CommonProps {
-    checkbox?: undefined;
-    switch?: undefined;
-    radioValue?: undefined;
-    fullPageOnWebView?: undefined;
-
+interface OnPressRowContentProps extends PressableProps, RightyRowProps, CommonProps {
     trackingEvent?: TrackingEvent | ReadonlyArray<TrackingEvent>;
     onPress: () => void;
```
Comment (atabel) [inline src/list.tsx:299]: If you are already extending `PressableProps`, is this `onPress` needed?

File: src/list.tsx [hunk 6]
```diff
@@ -258,107 +259,93 @@ type ControlProps = {
...
+function areCheckboxRowContentProps(obj: any): obj is CheckboxRowContentProps {
+    return typeof obj.checkbox !== 'undefined';
+}
+
+function areSwitchRowContentProps(obj: any): obj is SwitchRowContentProps {
+    return typeof obj.switch !== 'undefined';
+}
+
+function arePressableProps(obj: any): obj is PressableProps {
+    return obj.hasOwnProperty('onPress');
```
Comment (atabel) [inline src/list.tsx:325]: why do you check with `undefined` sometimes and with `hasOwnProperty` here?

File: src/list.tsx [hunk 7]
```diff
@@ -258,107 +259,93 @@ type ControlProps = {
...
+}
+
+function areOnPressRowContentProps(obj: any): obj is OnPressRowContentProps {
+    return (
+        arePressableProps(obj) &&
+        typeof obj.onPress !== 'undefined' &&
```
Comment (atabel) [inline src/list.tsx:331]: aren't these checks redundant?

File: src/list.tsx [hunk 8]
```diff
@@ -407,13 +394,18 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
     } = props;
 
     const radioContext = useRadioContext();
-    const disabled = props.disabled || (props.radioValue !== undefined && radioContext.disabled);
+    const disabled =
+        props.disabled || (areRadioRowContentProps(props) !== undefined && radioContext.disabled);
```
Comment (atabel) [inline src/list.tsx:398]: `areRadioRowContentProps` returns a boolean, why are you comparing with `undefined`?
shouldn't this be something like this?

```suggestion
        props.disabled || (areRadioRowContentProps(props) && radioContext.disabled);
```

File: src/list.tsx [hunk 9]
```diff
@@ -439,7 +431,11 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
             extra={extra}
             labelId={labelId}
             disabled={disabled}
-            withChevron={!!props.onPress || !!props.href || !!props.to}
+            withChevron={
+                (arePressableProps(props) && props.onPress !== undefined) ||
+                (areHrefRowContentProps(props) && props.href !== undefined) ||
+                (areToRowContentProps(props) && props.to !== undefined)
```
Comment (atabel) [inline src/list.tsx:437]: redundant checks again?

File: src/list.tsx [hunk 10]
```diff
@@ -517,8 +508,12 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
     }
 
     const renderRowWithControl = (Control: typeof Switch | typeof Checkbox) => {
-        const name = props.switch?.name ?? props.checkbox?.name ?? titleId;
-        return props.onPress ? (
+        const name =
+            ((areSwitchRowContentProps(props) && (props.switch?.name ?? undefined)) || undefined) ??
+            ((areCheckboxRowContentProps(props) && (props.checkbox?.name ?? undefined)) || undefined) ??
+            titleId;
```
Comment (atabel) [inline src/list.tsx:514]: why is this an improvement over the previous code? I find it quite more difficult to read

File: src/list.tsx [hunk 11]
```diff
@@ -604,7 +599,7 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
             className={classNames(classes.rowContent, classes.hover, classes.hoverDisabled)}
             role={role}
         >
-            {props.right
+            {areRightyRowProps(props) && props.right !== undefined
```
Comment (atabel) [inline src/list.tsx:602]: redundant checks?

File: src/list.tsx [hunk 12]
```diff
@@ -565,15 +559,15 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
         );
     };
 
-    if (props.switch) {
+    if (areSwitchRowContentProps(props)) {
         return renderRowWithControl(Switch);
     }
```
Comment (pladaria) [inline src/list.tsx:564]: are these helper functions really needed?

This is simpler and easier to read
```js
if (props.switch) {
    return ...
}
```

File: src/list.tsx [hunk 13]
```diff
@@ -648,14 +642,15 @@ interface HrefBoxedRowProps extends HrefRowContentProps, CommonBoxedRowProps {}
 interface ToBoxedRowProps extends ToRowContentProps, CommonBoxedRowProps {}
 interface OnPressBoxedRowProps extends OnPressRowContentProps, CommonBoxedRowProps {}
 
-type BoxedRowProps =
+type BoxedRowProps = ExclusifyUnion<
     | BasicBoxedRowProps
-    | SwitchBoxedRowProps
+    | (Omit<SwitchBoxedRowProps, 'switch'> & {switch: ControlProps | undefined})
```
Comment (pladaria) [inline src/list.tsx:647]: I don't understand this, why can't you use SwitchBoxedRowProps as is?

File: src/list.tsx [hunk 14]
```diff
@@ -439,7 +430,11 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
             extra={extra}
             labelId={labelId}
             disabled={disabled}
-            withChevron={!!props.onPress || !!props.href || !!props.to}
+            withChevron={
+                (arePressableProps(props) && props.onPress !== undefined) ||
```
Comment (pladaria) [inline src/list.tsx:434]: Is the previous condition no longer valid? 
```js
withChevron={!!props.onPress || !!props.href || !!props.to}
```

File: src/list.tsx [hunk 15]
```diff
@@ -258,107 +259,93 @@ type ControlProps = {
...
     | HrefRowContentProps
     | ToRowContentProps
-    | OnPressRowContentProps;
+>;
+
+function areRadioRowContentProps(obj: any): obj is RadioRowContentProps {
+    return typeof obj.radioValue !== 'undefined';
+}
+
+function areCheckboxRowContentProps(obj: any): obj is CheckboxRowContentProps {
```
Comment (Marcosld) [inline src/list.tsx:316]: Can `obj` be typed as `RowContentProps`?

**Review by Marcosld — APPROVED** [review]

File: src/list.tsx [hunk 16]
```diff
@@ -388,6 +338,18 @@ const useControlState = ({
...
+};
+
+const areCheckboxRowContentProps = (obj: any): obj is CheckboxRowContentProps => {
+    return 'checkbox' in obj;
+};
+
+const areRadioRowContentProps = (obj: any): obj is RadioRowContentProps => {
+    return 'radioValue' in obj;
+};
+
```
Comment (pladaria) [inline src/list.tsx:352]: consider this:
```diff
diff --git a/src/list.tsx b/src/list.tsx
index a2b9d543..bf342518 100644
--- a/src/list.tsx
+++ b/src/list.tsx
@@ -338,18 +338,6 @@ const useControlState = ({
     return [isChecked, toggle];
 };
 
-const areSwitchRowContentProps = (obj: any): obj is SwitchRowContentProps => {
-    return 'switch' in obj;
-};
-
-const areCheckboxRowContentProps = (obj: any): obj is CheckboxRowContentProps => {
-    return 'checkbox' in obj;
-};
-
-const areRadioRowContentProps = (obj: any): obj is RadioRowContentProps => {
-    return 'radioValue' in obj;
-};
-
 const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, ref) => {
     const titleId = useAriaId();
     const isInverse = useIsInverseVariant();
@@ -425,12 +413,7 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
         );
     };
 
-    if (
-        props.onPress &&
-        !areSwitchRowContentProps(props) &&
-        !areCheckboxRowContentProps(props) &&
-        !areRadioRowContentProps(props)
-    ) {
+    if (props.onPress && !props.switch && !props.checkbox && !props.radioValue) {
         return (
             <Touchable
                 ref={ref}
@@ -441,7 +424,7 @@ const RowContent = React.forwardRef<TouchableElement, RowContentProps>((props, r
                 dataAttributes={dataAttributes}
                 disabled={disabled}
             >
-                {renderTouchableContent(props)}
+                {renderTouchableContent(props as OnPressRowContentProps)}
             </Touchable>
         );
     }
```

**Review by pladaria — APPROVED** [review]

File: src/list.tsx (hunk 16 above)
Comment (brtbrt) [reply src/list.tsx:352]: told you in the meeting that it doesn't work! 😛
!()[https://media.giphy.com/media/l4pTsh45Dg7jnDM6Q/giphy.gif]

**Review by atabel — APPROVED** [review]

**Pull request merged**